### ISSUE TITLE
satisfiesExactlyInAnyOrder() fails if multiple items are equal as per equals() of items

### ISSUE THREAD

**Issue by jpgoelz** [assertj/assertj#3689]
**Describe the bug**
`satisfiesExactlyInAnyOrder()` fails when asserting multiple items that are equal according to their `equals()` method, even though they differ on the checked properties. At the same time, `satisfiesExactly()` passes with the same list.

One example:
We have classes that are only equal, if the ID field is equal. Other properties are not considered.
It is possible, that two items in a list share the same ID but differ in other properties.
When trying to check if it `satisfiesExactlyInAnyOrder()` using the other properties, it fails.

* assertj core version: 3.22.0
* java version: temurin-17.0.13
* test framework version: junit-jupiter 5.11.3
* os (if relevant): Win 10

**Test case reproducing the bug**

The class:
```java
public class MyObject {

    private long id;

    private String value;

   /* all args constructor, getters, setters */

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof final MyObject that)) {
            return false;
        }
        return getId() == that.getId();
    }

    @Override
    public int hashCode() {
        return Objects.hash(getId());
    }
}
```

The assertion:

```java
    @Test
    void myTest() {
        final List<MyObject> myObjects = List.of(
            new MyObject(0, "first"),
            new MyObject(0, "second")
        );

        assertThat(myObjects).satisfiesExactlyInAnyOrder(
            myObject -> assertThat(myObject.getValue()).isEqualTo("first"),
            myObject -> assertThat(myObject.getValue()).isEqualTo("second")
        );
    }
```

**Comment by scordio** [comment]
Hi @jpgoelz,

> * assertj core version: 3.22.0

You're using an almost three years old version 🙂 

Could you please upgrade to 3.26.3? I tried your example and it seems to work fine.

**Comment by jpgoelz** [comment]
How embarrassing. 🙈 We normally use [Renovate](https://github.com/renovatebot/renovate) to keep our dependencies up-to-date. So I expected AssertJ to be at the current version as well.
However, this is one of those rare cases where we did not specify the version and let Spring Boot (2.7.18) manage the dependency.
And so Renovate never suggested updates for it.

I can confirm that the tests are now passing, even with the real life code where I stumbled across the issue. So I am closing this bug report. 🎉
